torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp

```diff
@@ -116,6 +118,59 @@ static bool needsGradientInProfilingMode(Block* b) {
   return false;
 }
 
+bool guardDifferentiableGraph(Node* dnode) {
+  auto gi = dnode->g(attr::Subgraph)->inputs();
+  bool all_inputs_seen = true;
+  for (size_t i = 0; i < gi.size(); i++) {
+    auto ty = gi[i]->type()->cast<TensorType>();
+    if (ty) {
+      auto n = gi[i]->uses().at(0).user;
+      auto dni = dnode->inputs().at(i);
+      GRAPH_DEBUG("found first user of ", i, " as ", *n);
+      if (n->kind() == prim::profile) {
+        GRAPH_DEBUG(
+            "setting input ", i, " to type ", *n->ty(attr::profiled_type));
+        dni->setType(n->ty(attr::profiled_type));
+      } else if (dni->node()->kind() == prim::DifferentiableGraph) {
+        // The profiling node might have been absorbed in a preceding
+        // differentiable graph and thus not (not ideal for fusing either),
+        // see TestAutodiffSubgraphSlicing.test_does_not_create_cycles.
+        // Alternatives to this special casing could be specializing the types
+        // before autodiff or duplicating profile nodes for autodiff outputs
+        // but that should be done while creating subgraphs and would be
+        // a mess.
+        // XXX TODO: revisit the alternatives
+        Value* o = dni->node()->g(attr::Subgraph)->outputs().at(dni->offset());
+        if (o->node()->kind() == prim::profile) {
+          dni->setType(o->node()->ty(attr::profiled_type));
+        }
+      }
+
+      // we check if the optional is defined
+      all_inputs_seen &= (dni->type()->cast<TensorType>() != TensorType::get());
+    }
+  }
+  if (all_inputs_seen) {
+    // we may have seen both true and false for requires_grad. In this case
```

Comment:
Hmm, when would this happen exactly? I'm a little confused here. I guess this is the LSTM first loop doesn't require grad ?

Reply:
for LSTM in the first timestep the hidden state usually doesn't require gradients, the inputs would.